packages/engine/paima-funnel/src/funnels/midnight/funnel.ts

```diff
@@ -0,0 +1,451 @@
+      return baseData;
+    }
+
+    const result: {
+      blockNumber: number;
+      extensionDatums?: ChainDataExtensionDatum[];
+      internalEvents?: InternalEvent[];
+      submittedData?: SubmittedData[];
+    }[] = [];
+    let block = await this.waitForBlock(this.cache.nextBlockHeight);
+
+    for (const baseBlock of baseData) {
+      // Process all Midnight blocks that precede the base block.
+      // Break when we've seen a block in the relative future.
+      // TODO: apply confirmation depth here.
+      while (midnightTimestampToSeconds(block.timestamp) < baseBlock.timestamp) {
+        // The meat: process this one.
+        console.log(
+          `Midnight funnel ${this.chainInfo.config.networkId}: base #${baseBlock.blockNumber} / Midnight #${block.height}`
+        );
+
+        const extensionDatums = this.extensionsFromBlock(baseBlock.blockNumber, block);
+        if (extensionDatums.length) {
+          result.push({
+            blockNumber: baseBlock.blockNumber,
+            extensionDatums,
+          });
+        }
+
+        // Fetch next block.
+        this.cache.nextBlockHeight = block.height + 1;
+        block = await this.waitForBlock(this.cache.nextBlockHeight);
+      }
+    }
+
+    // Insert the internal MidnightLastBlock event into each ChainData.
+    const composed = composeChainData(baseData, result);
+    for (const chainData of composed) {
+      (chainData.internalEvents ??= []).push({
+        type: InternalEventType.MidnightLastBlock,
+        caip2: this.caip2,
+        block: this.cache.nextBlockHeight - 1,
+      });
+    }
+    return composed;
+  }
+
+  public override configPrint(): FunnelJson {
+    return {
+      type: 'MidnightFunnel',
+      chainName: this.chainInfo.name,
+      child: this.baseFunnel.configPrint(),
+    };
+  }
+}
+
+export class MidnightFunnelCacheEntry implements FunnelCacheEntry {
+  static readonly SYMBOL = Symbol('MidnightFunnelCacheEntry');
+
+  caip2: string;
+
+  /** Lowest Midnight block height that has not been processed by the funnel. */
+  nextBlockHeight: number = 0;
+
+  // WebSocket connection for being notified of new blocks.
+  wsClient: Client;
+  iter: AsyncIterableIterator<ExecutionResult<{ blocks: CachedBlock }, unknown>>;
+  wsBlocks: CachedBlock[];
+
+  constructor(config: MidnightConfig) {
+    this.caip2 = caip2PrefixFor(config);
+
+    this.wsClient = createClient({
+      url: config.indexerWS ?? defaultIndexerWs(config.indexer),
+      webSocketImpl: WebSocket,
+    });
+    this.iter = this.wsClient.iterate<{ blocks: CachedBlock }>({
+      query: `subscription { blocks { ${cachedBlockQuery} } }`,
+    });
+    this.wsBlocks = [];
+  }
+
+  async load(dbTx: PoolClient): Promise<MidnightFunnelCacheEntry> {
+    const checkpoint = await getMidnightCheckpoint.run({ caip2: this.caip2 }, dbTx);
+    console.log(checkpoint);
+    if (checkpoint.length > 0) {
+      this.nextBlockHeight = Number(checkpoint[0].block_height) + 1;
```

Comment:
will this be 0 if this branch is not executed? does this mean the funnel will always traverse the entire chain? in the other funnels we have some code in place so that we only start from the first block we need (either we do a binary search on the evm chains, or we just compute the slot if we can).